irohad/ametsuchi/tx_cache_outcome.hpp

```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_TX_CACHE_OUTCOME_HPP
+#define IROHA_TX_CACHE_OUTCOME_HPP
+
+#include <functional>
+
+#include <boost/variant.hpp>
+#include <cryptography/hash.hpp>
+
+#include "cryptography/hash.hpp"
+
+namespace iroha {
+  namespace ametsuchi {
+
+    /**
+     * Class represents the outcome of tx presence in a cache
+     */
+    class TxCacheOutcome {
+     public:
+      /// shortcut for tx hash type
+      using HashType = shared_model::crypto::Hash;
+
+     private:
+      /**
+       * Hash holder class
+       * The class is required only for avoiding duplication in outcome classes
+       */
+      class HashContainer {
+       public:
+        /// hash of the transaction
+        virtual const HashType &hash() const;
+
+        virtual ~HashContainer() = default;
```

Comment:
Polymorphism is not used in current implementation, so virtual methods look redundant. Is it expected to be used? If not, this class can be replaced by a struct with hash field, because TxCacheResponse  provides access by const reference, so no mutations are possible.

Same about inherited structs if this comment is applicable.